website/pages/docs/internals/limits.mdx

```diff
@@ -0,0 +1,271 @@
+one local secret engine (`cubbyhole`) and one auth engine mount (`token`).
+
+|   | Consul default (512 KiB) | Integrated storage default (1 MiB) |
+|---|--------------------------|------------------------------------|
+| Maximum number of namespaces | ~3500 | ~7000 |
+| Maximum number of namespaces with one additional secret engine per namespace | ~2300 | ~4600 |
+| Maximum nesting depth for namespaces | ~160 | ~220 |
+
+The maximum nesting depth calculation assumes a cost of 40 bytes per namespace
+path element.  160 nested paths = 160 namespaces ranging from 40 bytes to
+6400 bytes.
+
+The number of namespaces can be monitored by querying
+[`sys/namespaces`](/api-docs/system/namespaces).
+
+To estimate the number of namespaces that can be created, divide the mount
+point limit by the larger of of the number of auth mounts per namespace
+(including `ns_token`) and the number of secret mounts per namespace
+(including `identity` and `sys`.)
+
+### Entity and group limits
+
+The metadata that may be attached to an identity entity or an entity group
+has the following constraints:
+
+|            |      |
+|------------|------|
+| Number of key-value pairs in metadata |  64 |
+| Metadata key size | 128 bytes |
+| Metadata value size | 512 bytes |
+
+Vault shards the entities across 256 storage entries. This creates a
+hard limit of 128MiB storage space used for entities on Consul, or
+256MiB on integrated storage with its default settings. Entity aliases
+are stored inline in the Entity objects and so consume the same pool
+of storage. Entity definitions are compressed within each storage
+entry, and the pre-compression size varies with the number of entity
+aliases and the amount of metadata.  Minimally-populated entities
+about 200 bytes after compression.
+
+Group definitions are stored separately, in their own pool of 256
+storage entries. The size of each group object depends on the number
+of members and the amount of metadata.  Group aliases and group
+membership information is stored inline in each Group object. A group
+with no metadata, holding 10 entities, will use about 500 bytes per
+group.  A group holding 100 entities would instead consume about 4,000
+bytes.
+
+The following table shows a best-case estimate and a more conservative
+estimate for entities and groups. The number is slightly less than the
+amount that fits in one shard, to reflect the fact that the first
+shard to fill up will start inducing failures.  This maximum will
+decrease if each entity has a large amount of metadata, or if each
+group has a large number of members.
+
+|   | Consul default (512 KiB) | Integrated storage default (1 MiB) |
+|---|--------------------------|------------------------------------|
+| Maximum number of identity entities (best case, 200 bytes per entity) | ~610,000 | ~1,1250,000 |
+| Maximum number of identity entities (conservative case, 500 bytes per entity) | ~250,000 | ~480,000 | 
```

Comment:
I wonder if we should be more pessemistic for the conservative case, since some orgs seem to want to put a lot of metadata into entities?

Reply:
Maybe in addition to a conservative case, we also include an absolute worst case with maxed-out metadata for entities/groups? Since those are defined constants, it'll give customers an idea of how best-case, middle-ground case, worst-case differ from each other

Reply:
The numbers for the worst case (40960 bytes of incompressible metadata) are quite bad; I calculate them to be around 670 and 2400. 

The exact threshold depends on what you consider an acceptable probability of failure, as the buckets will fill up according to a binomial distribution.  But I don't have the right statistics math so I've been modelling them as independent, though this is of course incorrect.